src/apps/btc/btc_common.c

```diff
@@ -685,9 +686,42 @@ USE_RESULT bool btc_common_multisig_is_valid(
     return true;
 }
 
+// serialized_out must be of size BIP32_SERIALIZED_LEN.
+static bool _serialize_xpub(const XPub* xpub, uint8_t* serialized_out)
+{
+    struct ext_key wally_xpub __attribute__((__cleanup__(keystore_zero_xkey))) = {0};
+    if (!apps_common_bip32_xpub_from_protobuf(xpub, &wally_xpub)) {
+        return false;
+    }
+    if (bip32_key_serialize(
+            &wally_xpub, BIP32_FLAG_KEY_PUBLIC, serialized_out, BIP32_SERIALIZED_LEN) != WALLY_OK) {
+        return false;
+    }
+    return true;
+}
+
+static int _xpubs_sort_comp(const void* elem1, const void* elem2)
+{
+    const XPub* xpub1 = *(const XPub* const*)elem1;
+    const XPub* xpub2 = *(const XPub* const*)elem2;
+
+    uint8_t xpub1_serialized[BIP32_SERIALIZED_LEN] = {0};
+    uint8_t xpub2_serialized[BIP32_SERIALIZED_LEN] = {0};
+    if (!_serialize_xpub(xpub1, xpub1_serialized)) {
+        Abort("serialize_xpub");
+    }
+    if (!_serialize_xpub(xpub2, xpub2_serialized)) {
+        Abort("serialize_xpub");
+    }
+
+    // Sort by xpub serialization, ignoring the version bytes.
```

Comment:
I wonder why ignore the version. Shouldn't it always be the same? If it isn't, I would expect some check before this to fail the whole registration flow so it wouldn't even come to sorting.

This is what I see in external/.../wally_bip32.h:

```c
/** Version codes for extended keys */
#define BIP32_VER_MAIN_PUBLIC  0x0488B21E
#define BIP32_VER_MAIN_PRIVATE 0x0488ADE4
#define BIP32_VER_TEST_PUBLIC  0x043587CF
#define BIP32_VER_TEST_PRIVATE 0x04358394
```

Reply:
if it is mainnet or testnet is already encoded in the hash.

xpub versioning is very non-standard (Zpub, etc.), and to the device, it does not matter which is used by the computer wallet. if you use electrum (Zpub), then HWI (xpub), the multisig account is still the same and the user should not register a new one. in other words, different xpub versions should not change the hash.

Reply:
But those `define`s *are* the first 4 bytes according to https://github.com/digitalbitbox/libwally-core/blob/afd0542/src/bip32.c#L246-L255:

```c
    tmp32 = hdkey->version;
    if (!serialize_private) {
        /* Change version if serializing the public part of a private key */
        if (tmp32 == BIP32_VER_MAIN_PRIVATE)
            tmp32 = BIP32_VER_MAIN_PUBLIC;
        else if (tmp32 == BIP32_VER_TEST_PRIVATE)
            tmp32 = BIP32_VER_TEST_PUBLIC;
    }
    tmp32_be = cpu_to_be32(tmp32);
    out = copy_out(out, &tmp32_be, sizeof(tmp32_be));
```

in other words, `bip32_key_serialize` always encodes it in either `xpub` or `tpub` format?

Reply:
First of all "xpub" and "tpub" are only in the base58-check encoded string. The serialization we are looking at is the binary encoding of the xpub, where the first four bytes are one of the defines you posted. In our case hardcoded  to `BIP32_VER_MAIN_PUBLIC`.

But I discard them right after by hasing without the first four bytes.

Reply:
:+1: